### PR TITLE
Implement the read-write splitting plugin

### PR DESCRIPTION
**Pull request by aaron-congo**

### CODE REVIEW

**Review by sergiyvamz — APPROVED** [review]

**Comment by davecramer** [comment]
Since this is splitting on transaction boundaries, how are we handling read after writes ?

File: examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Properties;
+import software.amazon.jdbc.plugin.failover.FailoverFailedSQLException;
+import software.amazon.jdbc.plugin.failover.FailoverSuccessSQLException;
+import software.amazon.jdbc.plugin.failover.TransactionStateUnknownSQLException;
+import software.amazon.jdbc.plugin.readwritesplitting.ReadWriteSplittingPlugin;
+
+
```
Comment (joyc-bq) [inline examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java:33]: nit: extra space here

**Comment by congoamz** [comment]
@davecramer can you clarify your question? The plugin only switches at transaction boundaries if setReadOnly(true) has been called and reader load-balancing is enabled. Because of this, only reads can be performed in this state. The plugin does not switch between a writer/reader on transaction boundaries; instead, setReadOnly is used to switch

**Review by joyc-bq — APPROVED** [review]

**Comment by davecramer** [comment]
@congoamz thanks for the clarification. That makes sense. 
One thing to keep in mind is that opening connections on PostgreSQL is not free. I would expect that most apps will have a pool which minimizes the overhead, but if they don't then splitting on transaction boundaries might be an anti-pattern.

**Comment by congoamz** [comment]
@davecramer good to know, thanks. Switching to a new reader on transaction boundaries is disabled by default. We are working on creating some performance tests which will give us some insight into how this feature effects performance

**Pull request merged**